### PR TITLE
catppuccin-papirus-folders: init at unstable-2022-12-04

### PR DESCRIPTION
**Pull request by rubyowo**
###### Description of changes
Soothing pastel theme for Papirus Icon Theme folders. A Catppuccin port for https://github.com/PapirusDevelopmentTeam/papirus-folders.
Home Page: https://github.com/catppuccin/papirus-folders

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  version = "unstable-2022-12-04";
+
+  src = fetchFromGitHub {
+    owner = "catppuccin";
+    repo = "papirus-folders";
+    rev = "1a367642df9cf340770bd7097fbe85b9cea65bcb";
+    sha256 = "sha256-mFDfRVDA9WyriyFVzsI7iqmPopN56z54FvLkZDS2Dv8=";
+  };
+
+  buildInputs = [ bash getent gtk3 papirus-icon-theme ];
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:32]:
```suggestion
  nativeBuildInputs = [ gtk3 ];
```

The only build input that actually gets used is `gtk3` but it's used as a native build input for `gtk-update-icon-cache`

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+let
+  validAccents = [ "blue" "flamingo" "green" "lavender" "maroon" "mauve" "peach" "pink" "red" "rosewater" "sapphire" "sky" "teal" "yellow" ];
+  validFlavors = [ "latte" "frappe" "macchiato" "mocha" ];
+  pname = "catppuccin-papirus-folders";
+in
+lib.checkListOfEnum "${pname}: accent color" validAccents [ accent ]
+  lib.checkListOfEnum "${pname}: flavor"
+  validFlavors [ flavor ]
+  stdenvNoCC.mkDerivation
+rec {
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:21]:
```suggestion
lib.checkListOfEnum "${pname}: accent color" validAccents [ accent ]
lib.checkListOfEnum "${pname}: flavor" validFlavors [ flavor ]

stdenvNoCC.mkDerivation {
```

Indenting could be clearer (I stole this indenting from another random icon theme derivation which did something similar)

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    repo = "papirus-folders";
+    rev = "1a367642df9cf340770bd7097fbe85b9cea65bcb";
+    sha256 = "sha256-mFDfRVDA9WyriyFVzsI7iqmPopN56z54FvLkZDS2Dv8=";
+  };
+
+  buildInputs = [ bash getent gtk3 papirus-icon-theme ];
+
+  patchPhase = ''
+    substituteInPlace ./papirus-folders --replace "getent" "${getent}/bin/getent"
+  '';
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:36]:
```suggestion
  postPatch = ''
    patchShebangs ./papirus-folders
  '';
```

You only need to `patchShebangs` to get this script to run without bash and if you set `USER_HOME` and `DISABLE_UPDATE_ICON_CACHE`, it doesn't need `getent`

Also `postPatch` for that sort of thing is better than `patchPhase` (in case someone overrides with a particular patch file)

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  installPhase = ''
+    mkdir -p $out/share/icons
+    cp -r ${papirus-icon-theme}/share/icons/Papirus $out/share/icons
+    chmod -R u+rw $out
+    cp -r src/* $out/share/icons/Papirus
+    for theme in $out/share/icons/*; do
+        ${bash}/bin/bash ./papirus-folders -t $theme -o -C cat-${flavor}-${accent}
+        gtk-update-icon-cache --force $theme
+    done
+  '';
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:47]:
```suggestion
  installPhase = ''
    runHook preInstall

    mkdir -p $out/share/icons
    cp -r --no-preserve=mode ${papirus-icon-theme}/share/icons/Papirus $out/share/icons
    cp -r src/* $out/share/icons/Papirus
    for theme in $out/share/icons/*; do
        USER_HOME=$HOME DISABLE_UPDATE_ICON_CACHE=1 \
          ./papirus-folders -t $theme -o -C cat-${flavor}-${accent}
        gtk-update-icon-cache --force $theme
    done

    runHook postInstall
  '';
```

Notes:
* Using `--no-preserve=mode` on `cp` is clearer in intention than `chmod -R u+rw`
* You no longer need to start the script with explicit `bash` if you `patchShebangs` (although using something like that in `nativeBuildInputs` is better if it's something that `patchShebangs` won't help
* Set env vars to avoid `getent` usage (since that's pretty meaningless in the build sandbox anyway)
* Make sure when overriding a phase like `installPhase`, make sure to run the pre/post hooks too

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
+{ stdenvNoCC
+, lib
+, fetchFromGitHub
+, papirus-icon-theme
+, gtk3
+, bash
+, getent
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:7]:
```suggestion
```

These are no longer required

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -16125,4 +16125,10 @@
     github = "ziguana";
     githubId = 45833444;
   };
+  rubyowo = {
+    name = "Rei Star";
+    email = "perhaps-you-know@what-is.ml";
+    github = "rubyowo";
+    githubId = 105302757;
+  };
```
Comment (lilyinstarlight) [inline maintainers/maintainer-list.nix:16133]: This is buried deep within the manual (and we should add it to CONTRIBUTING.md too so it's more discoverable for new maintainers...), but additions to `maintainers/maintainter-list.nix` are supposed to be in their own commits titled "maintainers: add [name]": https://nixos.org/manual/nixpkgs/unstable/#var-meta-maintainers

**Review by lilyinstarlight — COMMENTED** [review]
This looks pretty good! I kinda want this now too since I plan to switch to catppuccin soon

I've put a few nits and changes below to have it follow a more consistent nixpkgs style and make it more overridable and such, but nothing too major

Edit: Oh and if you could also squash those top two commits that would be fabulous and make this merge-ready

File: maintainers/maintainer-list.nix (hunk 6 above)
Comment (rubyowo) [reply maintainers/maintainer-list.nix:16133]: ah okay thanks

**Comment by rubyowo** [comment]
wait oops

**Comment by rubyowo** [comment]
Ah, just noticed an issue that makes it so that only the icon theme Papirus gets copied, and not Papirus-Dark

**Review by lilyinstarlight — APPROVED** [review]
I could leave a few more style nits (like spaces around `[rubyowo]` or following indenting from other themes that use `checkListOfEnum` so that the derivation is unindented), but it's readable as-is and the code otherwise looks good now!

---

Result of `nixpkgs-review pr 207211` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>catppuccin-papirus-folders</li>
  </ul>
</details>

**Comment by rubyowo** [comment]
alejandra is not being very nice to me

File: pkgs/data/icons/catppuccin-papirus-folders/default.nix [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    postPatch = ''
+      patchShebangs ./papirus-folders
+    '';
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out/share/icons
+      cp -r --no-preserve=mode ${papirus-icon-theme}/share/icons/Papirus* $out/share/icons
+      cp -r src/* $out/share/icons/Papirus
+      cp -r src/* $out/share/icons/Papirus-Dark
```
Comment (lilyinstarlight) [inline pkgs/data/icons/catppuccin-papirus-folders/default.nix:40]: It looks like `Papirus-Dark` uses symbolic links in all of the places this would copy stuff to. So just copying this theme at all should be sufficient and you shouldn't need to duplicate it again (since it links to `../Papirus/[size]/places` anyway)

```suggestion
```

Comment (rubyowo) [reply pkgs/data/icons/catppuccin-papirus-folders/default.nix:40]: ah okay

Comment (lilyinstarlight) [reply pkgs/data/icons/catppuccin-papirus-folders/default.nix:40]: Good catch that it was missing Papirus-Dark though! That last push was just failing to build for me without removing this line, and it was printing errors like this:

```
catppuccin-papirus-folders-unstable> installing
catppuccin-papirus-folders-unstable> cp: cannot overwrite non-directory '/nix/store/crq0b2nv7skpi99lq33bjwc0n2y8fyzk-catppuccin-papirus-folders-unstable-2022-12-04/share/icons/Papirus-Dark/22x22/places' with directory 'src/22x22/places'
catppuccin-papirus-folders-unstable> cp: cannot overwrite non-directory '/nix/store/crq0b2nv7skpi99lq33bjwc0n2y8fyzk-catppuccin-papirus-folders-unstable-2022-12-04/share/icons/Papirus-Dark/24x24/places' with directory 'src/24x24/places'
catppuccin-papirus-folders-unstable> cp: cannot overwrite non-directory '/nix/store/crq0b2nv7skpi99lq33bjwc0n2y8fyzk-catppuccin-papirus-folders-unstable-2022-12-04/share/icons/Papirus-Dark/32x32' with directory 'src/32x32'
catppuccin-papirus-folders-unstable> cp: cannot overwrite non-directory '/nix/store/crq0b2nv7skpi99lq33bjwc0n2y8fyzk-catppuccin-papirus-folders-unstable-2022-12-04/share/icons/Papirus-Dark/48x48' with directory 'src/48x48'
catppuccin-papirus-folders-unstable> cp: cannot overwrite non-directory '/nix/store/crq0b2nv7skpi99lq33bjwc0n2y8fyzk-catppuccin-papirus-folders-unstable-2022-12-04/share/icons/Papirus-Dark/64x64' with directory 'src/64x64'
```

**Review by lilyinstarlight — APPROVED** [review]
This looks great to me now!

I tried to see if we could link from papirus-icon-theme (instead of copying from it) to save substituter space, but any scheme to do that would be much more complicated from what I can tell

So yeah let's stick with this implementation

---

Result of `nixpkgs-review pr 207211` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>catppuccin-papirus-folders</li>
  </ul>
</details>
